kge/job/train.py

```diff
@@ -530,37 +530,49 @@ def __init__(self, config, dataset, parent_job=None):
                 f(self)
 
     def _prepare(self):
-        from kge.indexing import prepare_index
-        # create sp and po label_coords (if not done before)
-        query_types = self.config.get("KvsAll.query_types")
-        self.used_query_types = [key for key, use in query_types.items() if use]
-        data = {}
-        self.queries = {}  # refers to actual queries, e.g. (AlbertEinstein, born_In)
-        self.labels = {}  # refers to the result(s) of an actual query
-        self.offsets = {}  # row in dictionary entry = starting offset in labels for corresponding query in queries
-        prev_end = 0
-        self.query_end_index = []  # entry gives end (of corresponding self.used_query_types entry) w.r.t. dataloader
-        for query_index, query_type in enumerate(self.used_query_types):
-            index_type = "sp_to_o" if query_type == "sp_" else ("so_to_p" if query_type == "s_o" else "po_to_s")
-            data[query_type] = self.dataset.index(f"{self.train_split}_{index_type}")
-            self.query_end_index.append(prev_end + len(data[query_type]))
-            prev_end = self.query_end_index[query_index]
-
-            # convert indexes to pytorch tensors: a nx2 queries tensor (rows = queries),
-            # an offset vector (row = starting offset in labels for corresponding
-            # query), a labels vector (entries correspond to values of original
-            # index)
-            #
-            # Afterwards, it holds:
-            # index[queries[i]] = labels[offsets[i]:offsets[i+1]]
+        from kge.indexing import index_KvsAll_to_torch
 
+        # determine enabled query types
+        self.query_types = [
+            key
+            for key, enabled in self.config.get("KvsAll.query_types").items()
+            if enabled
+        ]
+
+        #' for each query type: list of queries
+        self.queries = {}
+
+        #' for each query type: list of all labels (concatenated across queries)
+        self.labels = {}
+
+        #' for each query type: list of starting offset of labels in self.labels. The
+        #' labels for the i-th query of query_type are in labels[query_type] in range
+        #' label_offsets[query_type][i]:label_offsets[query_type][i+1]
+        self.label_offsets = {}
+
+        #' for each query type (ordered as in self.query_types), index of last example
+        #' of that type in the list of all examples
+        self.query_end_index = []
```

Comment:
Is it not the index of the last query of that type in the list of all queries (according to new definitions)?
Also, it is not the index of the last query but `1+ index of last query`, therefore the end in terms of pythonic slicing